src/Common/FileCache.cpp

```diff
@@ -598,6 +635,163 @@ bool LRUFileCache::tryReserve(
     return true;
 }
 
+LRUFileCache::ReserveResult LRUFileCache::tryReserveForQuery(const Key & key, size_t offset, size_t size, std::lock_guard<std::mutex> & cache_lock)
+{
+    if (!CurrentThread::isInitialized() || !CurrentThread::get().getQueryContext() || CurrentThread::getQueryId().size == 0)
+        return ReserveResult::NO_NEED;
+
+    auto query_id = CurrentThread::getQueryId().toString();
+    auto query_iter = query_map.find(query_id);
+
+    if (query_iter == query_map.end())
+        throw Exception(ErrorCodes::LOGICAL_ERROR, "Cache became inconsistent. There must be a bug");
+
+    auto query_context = query_iter->second;
+
+    /// The maximum cache capacity of the request is not reached, thus the
+    //// cache block is evicted from the main LRU queue by tryReserveForMainList().
+    if (query_context->cache_size + size < query_context->max_cache_size)
+    {
+        return ReserveResult::NO_NEED;
+    }
+    /// The maximum cache size of the query is reached, the cache will be
+    /// evicted from the history cache accessed by the current query.
+    else
+    {
+        size_t removed_size = 0;
+        size_t queue_size = queue.getElementsNum(cache_lock);
+
+        auto * cell_for_reserve = getCell(key, offset, cache_lock);
+
+        std::vector<FileSegmentCell *> trash;
+        std::vector<FileSegmentCell *> to_evict;
+        std::vector<LRUQueue::Iterator> ghost;
+
+        auto is_overflow = [&]
+        {
+            return (max_size != 0 && queue.getTotalWeight(cache_lock) + size - removed_size > max_size)
+            || (max_element_size != 0 && queue_size > max_element_size)
+            || (query_context->max_cache_size != 0 && query_context->cache_size + size - removed_size > query_context->max_cache_size);
+        };
+
+        /// Select the cache from the LRU queue held by query for expulsion.
+        for (auto iter = query_context->queue.begin(); iter != query_context->queue.end(); iter++)
+        {
+            if (!is_overflow())
+                break;
+
+            auto cell = getCell(iter->key, iter->offset, cache_lock);
+
+            if (!cell)
+            {
+                /// The cache corresponding to this record may be swapped out by
+                /// other queries, so it has become invalid.
```

Comment:
Good suggestion, in terms of implementation, we can find its query context based on the downloader_id of file_segment that will be reserved and update it. Is this correct?

Reply:
why need to update downloader id? we use this file segment in the scope of current query_id, so just get it the same way -- from `CurrentThread` and find needed context in the map by query id

Reply:
Sorry for not being clear, what I meant was that we may evict a file_segment that is not belong to the current query in tryReserveForMainList(). The query_context we actually want to update is the one that holds the file_segment (I think it's the downloader's request). But it may not be easy to find the query_context from file_segment, because each file_segment will clear its downloader_id after being downloaded. Therefore, I can only judge whether the cache in a query_context has been swapped out in tryReserveForQuery().